File: android/app/src/main/java/com/targetzone/library/ui/admin/UpiQrGenerator.kt
package com.targetzone.library.ui.admin

import android.graphics.Bitmap
import android.graphics.Color
import com.google.zxing.BarcodeFormat
import com.google.zxing.qrcode.QRCodeWriter
import com.targetzone.library.data.api.BASE_URL
import java.net.URLEncoder

// Mirrors frontend/src/utils/upiPay.js — kept in sync manually, no shared
// codegen between the web and Android clients here.
private fun enc(s: String) = URLEncoder.encode(s, "UTF-8")

fun buildUpiDeepLink(vpa: String, payeeName: String, amount: Double, note: String): String {
    val amountStr = "%.2f".format(amount)
    return "upi://pay?pa=${enc(vpa)}&pn=${enc(payeeName)}&am=${enc(amountStr)}&cu=INR&tn=${enc(note)}"
}

fun buildPayRedirectLink(vpa: String, payeeName: String, amount: Double, note: String): String {
    val origin = BASE_URL.removeSuffix("api/").removeSuffix("/")
    val amountStr = "%.2f".format(amount)
    return "$origin/pay?pa=${enc(vpa)}&pn=${enc(payeeName)}&am=${enc(amountStr)}&cu=INR&tn=${enc(note)}"
}

// Encode-only — no scanning/camera dependency, just com.google.zxing:core.
fun generateQrBitmap(content: String, sizePx: Int = 512): Bitmap {
    val writer = QRCodeWriter()
    val matrix = writer.encode(content, BarcodeFormat.QR_CODE, sizePx, sizePx)
    val bmp = Bitmap.createBitmap(sizePx, sizePx, Bitmap.Config.RGB_565)
    for (x in 0 until sizePx) {
        for (y in 0 until sizePx) {
            bmp.setPixel(x, y, if (matrix.get(x, y)) Color.BLACK else Color.WHITE)
        }
    }
    return bmp
}

File: ios/LibrarySystem/LibrarySystem/Features/Admin/UpiQrGenerator.swift
import SwiftUI
import CoreImage.CIFilterBuiltins

// Mirrors frontend/src/utils/upiPay.js and android/.../UpiQrGenerator.kt —
// kept in sync manually, no shared codegen between the three clients here.
// Uses Core Image's built-in QR generator (part of the standard iOS SDK —
// no third-party dependency needed).
func buildUpiDeepLink(vpa: String, payeeName: String, amount: Double, note: String) -> String {
    var components = URLComponents()
    components.scheme = "upi"
    components.host = "pay"
    components.queryItems = [
        URLQueryItem(name: "pa", value: vpa),
        URLQueryItem(name: "pn", value: payeeName),
        URLQueryItem(name: "am", value: String(format: "%.2f", amount)),
        URLQueryItem(name: "cu", value: "INR"),
        URLQueryItem(name: "tn", value: note),
    ]
    return components.url?.absoluteString ?? "upi://pay"
}

func generateQrImage(from string: String, sizePt: CGFloat = 240) -> UIImage {
    let context = CIContext()
    let filter = CIFilter.qrCodeGenerator()
    filter.message = Data(string.utf8)
    filter.correctionLevel = "M"

    guard let outputImage = filter.outputImage else {
        return UIImage(systemName: "qrcode") ?? UIImage()
    }
    // The raw CI output is only a few pixels across — scale up before
    // rendering so it isn't blurry at display size.
    let scale = sizePt / outputImage.extent.width
    let scaled = outputImage.transformed(by: CGAffineTransform(scaleX: scale, y: scale))

    guard let cgImage = context.createCGImage(scaled, from: scaled.extent) else {
        return UIImage(systemName: "qrcode") ?? UIImage()
    }
    return UIImage(cgImage: cgImage)
}
